pdns/tcpreceiver.cc

```diff
@@ -773,21 +773,33 @@ int TCPNameserver::doAXFR(const DNSName &target, std::unique_ptr<DNSPacket>& q,
 
   // Catalog zone start
   if (di.kind == DomainInfo::Producer) {
-    // Ignore all records except NS at apex
-    sd.db->lookup(QType::NS, target, di.id);
+    const DNSName customPropertiesSuffix = DNSName("ext") + target;
+
+    /* make sure the record containing the version is added first, in case the consumer is picky */
+    zrrs.insert(zrrs.begin(), CatalogInfo::getCatalogVersionRecord(target));
+
+    sd.db->list(target, di.id);
     while (sd.db->get(zrr)) {
-      zrrs.emplace_back(zrr);
+      // Ignore all records except:
+      // 1. NS at apex
+      if (zrr.dr.d_name == target && zrr.dr.d_type == QType::NS) {
+        zrrs.emplace_back(zrr);
+      }
+      // 2. Custom properties
+      else if (zrr.dr.d_name.isPartOf(customPropertiesSuffix)) {
+        zrrs.emplace_back(zrr);
+      }
     }
-    if (zrrs.empty()) {
+
+    /* if we only have the version record, synthesize a NS record */
+    if (zrrs.size() == 1) {
```

Comment:
Detect ns records in the loop above and only add the default one when there are no NS records present in the zone.